### ISSUE TITLE
Feature request - have a way to put back preset of filter/orders/pagination

### ISSUE THREAD

**Issue by ghiscoding** [ghiscoding/aurelia-slickgrid#14]
**I'm submitting a feature request**

* **Browser:**
all

* **Language:**
TypeScript

**Current behavior:**
not implemented

* **What is the motivation / use case for changing the behavior?**
I would like to be able to load the grid with some presets (filters/orders/pagination) and/or expose them so that we can save them in LocalStorage or reload them from LocalStorage/DB.

**Comment by ghiscoding** [comment]
@jmzagorski 
I opened for references and to not mix too many subjects in the same thread.
It is a continuation of what we started to talk [here](https://github.com/ghiscoding/aurelia-slickgrid/pull/13#issuecomment-368242379)

and my reply to your post is the following:

Yeah we will use LocalStorage to keep filters/sorters/pagination when coming back on the page. Basically we have list and if open for detail (which is a separate window), the user wants to see the exact same grid with his last filters/sorters. It's similar to do a "back" and put back the grid in the same state. I got the preset of the filters working, I still have to do the other ones. 

What you talked about with your old grid is something different but I would also like to do that as well in the future. I'm just not sure if it would go in the same structure for these hiding/showing of certain columns.Tell me what you think and I could modify the structure if need be (I'm still working on it with my Angular repo). Here's the current structure I have (filters is working)

Preset defined in the grid options
```javascript
    this.gridOptions = {
      enableFiltering: true,
      presets: {
        // you can also type operator as string, e.g.: operator: 'EQ'
        filters: [
          { columnId: 'gender', searchTerm: 'male', operator: OperatorType.equal },
          { columnId: 'name', searchTerm: 'John Doe', operator: OperatorType.contains },
          { columnId: 'company', searchTerms: ['xyz'], operator: 'IN' }
        ],
        sorters: [
          { columnId: 'name', direction: SortDirection.asc },
          { columnId: 'company', direction: SortDirection.desc }
        ],
        pagination: { pageNumber: 2 }
      },
```
These are not in the grid definition, it's make it easier for the user to simply write it in there (or use LocalStorage to dump it back).

**Comment by jmzagorski** [comment]
I was inspired to update an old github repo of slick plugins from when I first coded in javascript to see the structure I used for storing my columns in localStorage. You will see [here](https://github.com/jmzagorski/SlickGridExtensions/blob/master/slick.ext.columnstore.js) i stored the `columnDefinition` every time the `onColumnsReordered` was triggered because that is how the [Slick.Controls.ColumnPicker](https://github.com/mleibman/SlickGrid/blob/bf4705a96c40fea088216034def4d0256a335e65/controls/slick.columnpicker.js) updated itself, but that was problematic with any functions because those dont get stored in localStorage as you can see [here](https://github.com/jmzagorski/SlickGridExtensions/blob/master/slick.ext.columnstore.js#L36). 

A few question to help me better understand your presets. I tried to get an idea of you logic by looking at you Angular code:

1. Would you store the `gridOptions` in localStorage/DB based on the `grid-id` so you know which `gridOptions` to fetch if there are multiple grids?
2.  I was looking at https://github.com/ghiscoding/Angular-Slickgrid/commit/6dc8875f224c24b59ea3ccf5082801feddef1222#diff-8d207e24758c47914b90b634a07bc2d7R372 and it is mapping the column preset from the `gridOption` to the `columnDefinition`.  You said used the `gridOptions` to store the presets over the `columnDefinition`

> so it makes it easier for the user to simply write them in there (or use a dump from LocalStorage/DB)

I guess I did not understand that sentence (sorry I am still learning your code)? Can the `columnDefinitions` not be written to as easily? l . Is one benefit that the filter can go on any grid that has at least a `columnDefinition.id` that matches a filter? Thanks for helping me understand.

**Comment by ghiscoding** [comment]
To answer back
1. I actually didn't do that part yet (save in LocalStorage) but the idea is to save the `presets` object into LocalStorage and when going on that page re-use that `presets` if found in LocalStorage or DB (user preference). The order would be
    1. Do we have presets in LocalStorage? Yes use it to load grid, else continue to ii)
    2. Do we have user preference presets? Yes use it to load grid, else continue to iii)
    3. No presets found, load the grid regularly
2. I would save the current grid filters/orders before the grid gets destroyed, as shown [here](https://github.com/ghiscoding/Angular-Slickgrid/blob/6dc8875f224c24b59ea3ccf5082801feddef1222/src/app/examples/grid-odata.component.html#L26) and [here](https://github.com/ghiscoding/Angular-Slickgrid/blob/6dc8875f224c24b59ea3ccf5082801feddef1222/src/app/examples/grid-graphql.component.ts#L169). I will probably have to build a `presets` object myself by merging current filters/orders into a presets object, not a big deal.
3. **Note 1** The Filter `searchTerm(s)` still works, I actually use that code in each Filter to do the DOM element pre-populate (this way I didn't have to recode anything in any of the filters). However if `presets` are defined, it will use that instead of Filter searchTerm(s), so the order is this:
    1. Do we have presets? Yes use them, disregard Filter searchTerm(s), else go to ii)
    2. Do we have Filter searchTerm(s)? Yes use them to pre-populate header row filters
    3. No presets/searchTerm(s). load grid regularly without any filters/orders
4. **Note 2** the latest version of SlickGrid [version `2.3.13`](https://github.com/6pac/SlickGrid/releases/tag/2.3.13) brought in a new Slick Event `onColumnsChanged` in the ColumnPicker/GridMenu which returns an array of the currently shown columns. That will probably help you with the column store that you had implemented.

So at the end, I would like to incorporate, if possible, the column store within the `presets` structure/object. I'm still working on the Angular version of this and if there's any structure change to do in order to get also the column store in the game, then let me know. I'm also wondering if the word `presets` is a good property name, though I don't know what else to use (BTW, English is my 2nd language, so you might see typos or strange words used here and there lol)

**Comment by jmzagorski** [comment]
Thanks. English is my first and only language, but I am definitely no expert lol. I will try to help anywhere I see typos or strange words, but I might do that as well haha.

Looking at my `ColumnStore`, the purpose of it was to keep track of which columns were hidden and their exact order if the user had moved columns. The plugin caches the columnDefinitions directly, which is different than the `gridOptions` with your presets. 

Not 100% sure if this would work, but here is a possible solution . To use the `gridOptions` you will have to save the `id` and the current `index` of the columns during `onGridDestroy` by calling `grid.getColumns`. `getColumns` will return only visible columns and it have the columns in the correct order if the user reorders their columns. So you could have a `columnDefinition` property on your presets such as the example below. Then, to restore the columns you call `grid.getColumns` when the grid intiailizes, iterate over those original columns, remove/reorder the original columns and then call `grid.setColumns`, Hopefully, that makes sense, if not I can provide a more detailed explanation

```javascript
this.gridOptions = {
      presets: {
        // you can also type operator as string, e.g.: operator: 'EQ'
        filters: [
          { columnId: 'gender', searchTerm: 'male', operator: OperatorType.equal },
          { columnId: 'name', searchTerm: 'John Doe', operator: OperatorType.contains },
          { columnId: 'company', searchTerms: ['xyz'], operator: 'IN' }
        ],
        sorters: [
          { columnId: 'name', direction: SortDirection.asc },
          { columnId: 'company', direction: SortDirection.desc }
        ],
        pagination: { pageNumber: 2 },
       columnDefinitions: [
         // in this example gender is hidden because it is not here and company was moved to the first column
         // because the index is 0
         { columnId: 'name', index: 1 },
         { columnId: 'company', index: 0 },
       ]
      }, 
```

**Comment by ghiscoding** [comment]
I think I'll have to re-read that twice to understand the `index: 1` in your example but at first look, I would rename `columnDefinitions` to simply `columns` to remove confusion with the real `columnDefinitions` defined earlier.

**Comment by jmzagorski** [comment]
Lets see if I can better explain: the index is the column index in the columns array. For example, you start with these two columns for your grid:

```javascript
this.columns = [{
  id: 'gender',
  name: 'Gender',
  field: 'gender' 
}, {
  id: 'company',
  name: 'company',
  field: 'company'
}];
```

`Gender` has an index of 0 and `company` has an index of 1. If the user drags the `company` column in front of gender column, and you call `grid.getColumns`, you will get this:

```javascript
this.columns = [{
  id: 'company',
  name: 'company',
  field: 'company'
}, {
  id: 'gender',
  name: 'Gender',
  field: 'gender' 
}];
```

`company` will be the first column in the list (index 0). In order to save the column order you need to know that the user has placed the `company` column in front of the `gender` column. When you recreate the grid from presets, the grid will be initialized with `gender` in front of `company` (index 0) as in the first example. However, your presets will show that `company` should be at index 0. So you need to recreate the columns array with `company` at index 0 and then call `grid.setColumns`

**Comment by ghiscoding** [comment]
Hmm ok I understand your index now, when I read it the first time I was about to leave to play sport and I was in a hurry. Anyhow, I think you can simplify it though. The sample that I gave you has the sorters (which I didn't implement yet) and the way I want it to work is that the position inside the array will define what column will be sorted first, I do exactly the same in our [GraphQL - orderBy](https://github.com/ghiscoding/aurelia-slickgrid/wiki/GraphQL-Sorting#orderby) implementation which is also easy to code since you just have to loop in sequence. 

The structure I had in mind for these columns would look more like this
```javascript
this.gridOptions = {
      presets: {
        // you can also type operator as string, e.g.: operator: 'EQ'
        filters: [
          { columnId: 'gender', searchTerm: 'male', operator: OperatorType.equal },
          { columnId: 'name', searchTerm: 'John Doe', operator: OperatorType.contains },
          { columnId: 'company', searchTerms: ['xyz'], operator: 'IN' }
        ],
        sorters: [
          // the orderBy is first by name ASC then by company DESC
          { columnId: 'name', direction: SortDirection.asc },
          { columnId: 'company', direction: SortDirection.desc }
        ],
        pagination: { pageNumber: 2 },
       columns: [
         // in this example gender and the columns are in this order: name, gender, company
         { columnId: 'name', show: true },
         { columnId: 'gender', show: false},
         { columnId: 'company', show: true },
       ]
      }, 
```

isn't it easier to read and work which such structure? I don't think users want to start counting indexes and also in your example, you would have to sort your array by index before doing anything. The goal is to make it simple and easy to understand for the user, we also want to store that in DB/LocalStorage easily so that we can save/reload these presets. 

The only problem we might have is that the ColumnPicker/GridMenu with this new Slick Event `onColumnsChanged` only returns the ones that are shown on the screen, in this case we lose the hidden ones and it's not convenient. For that we might have to loop through both `for` loops to compare which are ones are shown/hidden.

**Comment by ghiscoding** [comment]
Oh wait, I think I now understand why you have that index and why the columns were in that order. So you were basically re-using the same order as the `columnDefinitions` array, is that it? That's actually less `for` loop compare to my propositions. It's actually something to take in consideration for performance reasons. 

Another thing that we could do, similar to your index, with the new Export to File that I am currently working on, the Grid Menu is starting to be crowded, so I added a new property called `positionOrder`, it's basically the same as your index but it 1 based (not 0 based). A lower number will make it appear the command show first in the Grid Menu list. So yeah almost the same as your index. 

Anyhow, I still find it easier to just go with the position in the array representing the position to show in the UI. Because if we follow the `columnDefinitions` position, like you said with the index, and the next day the Dev code changes and add 1 column in the middle of that definition then your `columns: [ ]` array is out of sync, while if we follow it like I said, the only thing that will happen is that the column might not show up because it's not yet part of the saved `columns: [ ]`

**Comment by jmzagorski** [comment]
Yes, my old code was using the same order as the columnDefinition so it was a fast and easy way to get the grid back to the state the user had it in originally. Good idea about `positionOrder`, I think that will work nicely.

**Comment by ghiscoding** [comment]
@jmzagorski 
I'm looking at coding a function to return all current filters/sorters/pagination and column order in the future. I'm just unsure on how to name the function, I can't think of anything else but `getCurrentGridState()` but I wonder if the word `State` might be confused for something else like the upcoming Aurelia State. Any other suggestions?

**Comment by jmzagorski** [comment]
hmm naming is one of the greatest challenges in programming. I'll think about it today and get back to you in this comment with an EDIT

**Comment by ghiscoding** [comment]
So far a grid option preset looks like this:
```javascript

this.gridOptions = {
  presets: {
    filters: [
      { columnId: 'gender', searchTerm: 'female' },
      { columnId: 'company', searchTerms: ['abc', 'xyz'] }
    ],
    sorters: [
      { columnId: 'firstName', direction: SortDirection.ASC },
      { columnId: 'lastName', direction: SortDirection.DESC }
    ],
    pagination: { pageNumber: 2, pageSize: 20 },

    // possibly in the future
    columns: [ ],
    grouping: [ ]
  }
};
```

**Comment by jmzagorski** [comment]
since you are getting presets, `getGridPresets`, `getGridCurrentPresets`, `getGridOptionsPresets`, `getSavedPresets`, `getPresetState`, `getCurrentPresetState`. Your `getCurrentGridState` is good too, but then you might have to define what you mean by `state`, since state can be the actual data on the grid too. Also you mention `state` in`Editor.applyValue: (item: any, state: any) => void;`

**Comment by ghiscoding** [comment]
The `Editor.applyValue: (item: any, state: any) => void;` is the structure defined by SlickGrid, not by me.

The idea is that the `getCurrentGridState()` will call internally `getCurrentFilters()`, `getCurrentSorters()`... and build a `presets` object that will be returned to the user. This avoid the users having to create the `presets` object by himself. For example, the Backend Services have no idea or concept about `presets`, they deal with filter/sorter/pagination changed, it doesn't have to know the `presets` at all, that is the job of the `aurelia-slickgrid` controller, that controller will have the `getCurrentGridState()` and pull the necessary  `getCurrentFilters()`, `getCurrentSorters()` from the backend service (whichever that is) or the local ones (that is without a backend service). 

Not sure if I explain it well. But anyhow, I can't just call it `getGridCurrentPresets` because I want to know their state when I leave the page and save that in LocalStorage for example. So again how I want to use it is following this:
1. Do I have some presets saved in LocalStorage? Yes use them, else go to step 2
2. Do I have user preference presets saved in DB? Yes use them, else go to step 3
3. No presets (empty object), load grid as normal by following the column definition and grid definition

**Comment by jmzagorski** [comment]
> The Editor.applyValue: (item: any, state: any) => void; is the structure defined by SlickGrid, not by me.

I know that value is defined by SlickGrid, I was just making the point that the word `state` is used in that context too

Here is my opinion on naming conventions for methods: what is the ultimate purpose of the method? From your last reply it appears to be:

> build a presets object that will be returned to the user

`getCurrentGridState()` will call **internally** `getCurrentFilters()` and `getCurrentSorters()` and maybe some other methods either using a backend service or a local service. Whichever service it uses to `getCurrentFilters()` and `getCurrentSorters()` is your internal logic on how to build the preset object. When I use the `getCurrentGridState()` method, I am not concerned about "how are you getting my presets" (using LocalStorage or a DB),  I just care that "you are getting my presets". IMHO `getCurrentGridState()` might be a little vague because I could ask myself "what state are you returning to me when I call this function?" Again, if I understand you correctly, you are returning my preset object, which is why all my examples have the word `preset` in it.

Hopefully that makes sense since I did not really explain that in my prior post.

**Comment by ghiscoding** [comment]
hmm yeah good point, the end result is mostly to build an object of the type Preset indeed. However it could be used to simply know what are the current state of each filters/sorters/pagination/... at that point in time. 

Is the name `GridStateService` any good or do you have another proposition?

**Comment by jmzagorski** [comment]
I think `GridStateService` is fine if that is what you had in mind, and if that service is to return different types of state in the future. Would this `GridStateService` have the method to build the presets object?

**Comment by ghiscoding** [comment]
Yes it will, or perhaps I could change the type from `Preset` to `GridState` and make the grid option `presets` property be of type `GridState`.

This might be better to tell the user that he can get any of these state (or all states) from that Service and it's not just for build a preset, it's good to know what do we have currently set? 

Is that better or even more confusing? 
The user will use that Service in his ViewModel, the structure could look like this:

Option 1
```javascript
export interface GridOtions {
  /** Query presets before grid load (filters, sorters, pagination, ...) */
  presets?: GridState;
}

export interface GridState {
  filters: CurrentFilter[];
  sorters: CurrentSorters[];
  pagination: { pageNumber: number; pageSize: number; }
  columns: CurrentColumns[];
  groups: CurrentGroups[];
}

export class GridStateService {

  getCurrentFilters(): CurrentFilter[] { }

  getCurrentSorters(): CurrentSorter[] { }

  // this method will return an object that is the same structure of what will be used to build "presets"
  getCurrentGridStates(): CurrentGridState { }
}
```

or I change everything back to have the word Preset in all Types

Option2
```javascript
export interface GridOtions {
  /** Query presets before grid load (filters, sorters, pagination, ...) */
  presets?: Preset;
}
export interface GridState {
  filters: PresetFilter[];
  sorters: PresetSorters[];
  pagination: { pageNumber: number; pageSize: number; }
  columns: PresetColumns[];
  groups: PresetGroups[];
}

export class GridStateService {

  getCurrentFilters(): PresetFilter[] { }

  getCurrentSorters(): PresetSorter[] { }

  // this method will return an object that is the same structure of what will be used to build "presets"
  getCurrentGridStates(): Preset { }
}
```

What do you prefer? I'm a bit for option 1.

**Comment by jmzagorski** [comment]
Yes option 1 because of what you said about "getting any state" . You can always refactor later if its not used as you thought it would be.

**Comment by ghiscoding** [comment]
Thanks, it really helps in discussing. Don't want to be stuck with bad naming and bad assumptions. :)

**Comment by jmzagorski** [comment]
Very true. I am trying to fix bad naming on my application right now.

**Comment by ghiscoding** [comment]
Practically done with all of this in my Angular repo, I got it working with BackendServiceApi and without which I call regular local grid. The only difference in the `GridState` is that the Pagination property will be empty (perhaps I should remove that property from local grid) on regular non-backend grid.

From the print screen below, you can see the top portion is what gets loaded by the "presets" and the console at the bottom is what shows up when calling `this.gridStateService.getCurrentGridState()` the `onBeforeGridDestroy` event and I would save to LocalStorage at that time. 

So it look like I can merge my Angular branch soon and expect to have it by next week in Aurelia :)
Also, we can add `columns` and `groups` in the future. I'd be happy to have a PR on the `columns` as I don't think that I will work on that yet in Angular, got to put the grid aside for a while, it's mainly enough features for what we need.

The grid option with the "presets" is defined as followed:
```javascript
this.gridOptions = {
      autoResize: {
        containerId: 'demo-container',
        sidePadding: 15
      },
      enableFiltering: true,
      
      // use columnDef searchTerms OR use presets as shown below
      presets: {
        filters: [
          { columnId: 'duration', searchTerms: [2, 22, 44] },
          { columnId: 'complete', searchTerm: '>5' }
        ],
        sorters: [
          { columnId: 'duration', direction: 'DESC' },
          { columnId: 'complete', direction: 'ASC' }
        ],
      }
    };
```


![preset](https://user-images.githubusercontent.com/643976/36817393-0531e470-1caf-11e8-8110-ecaffac01785.png)

**Comment by jmzagorski** [comment]
Great work!. I don't have a need yet for the columns or groups, but our "dashboard", which has slickgrid, is very new in our app, and if users want to hide/show columns I will work on the PR in the future.

**Comment by ghiscoding** [comment]
hey I'd be curious to know what are the options that you're using (with/without backend service, if so which one). From what I can tell with a few of our discussions, you're using regular local grid (non-backend) and want to use inline Editors with I guess a few Formatters, some Filters too I assume.

And are you able to do a print screen, you can delete it after few minutes if there's any sensible data shown. I'm just curious to see what my lib is used with and what are the best options of the lib that you like. Just to see if my efforts are aligned in roughly the same areas or not. 

I'm starting to see some traction in both repos (Aurelia mainly from you and Angular where I got couple of PRs and good feedback).

**Comment by jmzagorski** [comment]
I will be happy to get some info together for you since many of the grid don't have sensitive data

I will respond in https://github.com/ghiscoding/aurelia-slickgrid/issues/10 so it stays with the documentation conversation.

**Comment by ghiscoding** [comment]
Closed by PR #23
